074-connector-offsets-support.md

```diff
@@ -0,0 +1,180 @@
+# Connector Offsets Support
+
+Update the KafkaConnector resource to allow users to manage offsets
+
+## Current situation
+
+[KIP-875][kip] added first class offsets support to Kafka Connect.
+This was done via three new endpoints:
+* GET /connectors/{connector}/offsets
+* PATCH /connectors/{connector}/offsets
+* DELETE /connectors/{connector}/offsets
+* PUT /connectors/{connector}/stop
+
+Support for the "stop" endpoint was added via [PR 9095](https://github.com/strimzi/strimzi-kafka-operator/pull/9095).
+Currently users that are utilising the KafkaConnector resource cannot make use of the API endpoints for managing offsets.
+
+## Motivation
+
+Previously if users wanted to list, update or delete offsets for a connector they had to interact with the relevant Kafka topics directly.
+[KIP-875][kip] adds new endpoints to let the user do this more easily.
+There are multiple reasons a user might want to manage offsets, for example (as listed in the KIP):
+* Resetting the offsets for a connector while iterating rapidly in a development environment (so that the connector does not have to be renamed each time)
+* Viewing the in-memory offsets for source connectors on the cluster in order to recover from accidental deletion of the Connect source offsets topic (this is currently possible with the config topic, and can be a life-saver)
+* Monitoring the progress of running connectors on a per-source-partition basis 
+* Skipping records that cause issues with a connector and cannot be addressed using existing error-handling features
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaConnector resources.
+This proposal suggests different approaches for each action, since they are all slightly different.
+In summary:
+ - listing offsets via the KafkaConnector status and enabled/disabled through a property
+ - altering offsets via a new resource called KafkaConnectorOffsets
+ - deleting offsets via an annotation on the KafkaConnector resource
```

Comment:
I hadn't, but presumably it wouldn't be hard to implement, it would just require additional steps for the operator to follow.

For deleting offsets when deleting a connector, I guess the operator would have to do the following steps:
1. stop the connector
2. reset offsets
3. delete the connector

I think we would want a way to disable it, but I wonder if it should be the default behaviour since you can now stop connectors, I'm not sure why you would delete it if you wanted the offsets to be maintained.

For specifying offsets when creating a connector the steps would be:
1. create the connector in a stopped state (we don't currently support this as far as I know but Connect has made it possible)
2. alter the offsets
3. resume the connector
With the current proposal I'm not sure how this would fit in, it would perhaps work better to have the user do it. Although it could be an argument for not having the separate resource, then it would be easier for the operator to do everything.

What do you think?

Reply:
Once you delete the connector, you can't delete, change or list them via the `/offsets` endpoint anymore. You can still do it by directly interacting with the either the consumer group for sink connectors, or the topic for source connectors.

Reply:
Resolving conversation, since I've added this as an option now